objectbox/lib/src/native/store.dart

```diff
@@ -272,8 +273,49 @@ class Store {
   /// boxes. In addition, you get a consistent (transactional) view on your data
   /// while the transaction is in progress.
   @pragma('vm:prefer-inline')
-  R runInTransaction<R>(TxMode mode, R Function() fn) =>
-      Transaction.execute(this, mode, fn);
+  R runInTransaction<R>(TxMode mode, R Function() fn) {
+    // Whether the function is an `async` function. We can't allow those because
+    // the isolate could be transferred to another thread during execution.
+    // Checking the return value seems like the only thing we can in Dart v2.12.
+    if (fn is Future Function() && _nullSafetyEnabled) {
+      // This is a special case when the given function always throws. Triggered
+      //  in our test code. No need to even start a DB transaction in that case.
+      if (fn is Never Function()) {
+        // WARNING: don't be tempted to just `return fn();` - the code may
+        // execute DB operations which wouldn't be rolled back after the throw.
+        throw UnsupportedError('Given transaction callback always fails.');
+      }
+      throw UnsupportedError(
+          'Executing an "async" function in a transaction is not allowed.');
+    }
+
+    return _runInTransaction(mode, (tx) => fn());
+  }
+
+  /// Internal only - bypasses the main checks for async functions, you may
+  /// only pass synchronous callbacks!
+  R _runInTransaction<R>(TxMode mode, R Function(Transaction) fn) {
```

Comment:
What happens if the `mode` changes? Should this function check for that?

Reply:
Hm, maybe not. The database should throw if the mode isn't correct.

Reply:
yep, core does

Reply:
hmm, actually there was a check for a reused Tx but it got lost in this final version

Reply:
added